knox: handle race condition on concurrent logout call

xrmx opened the pull request:
With AUTO_REFRESH enabled authentication is racing against token removal of logout. If a token gets removed updating its expiry would led to a DatabaseError raised by the Django ORM.
    
Fix that by ignoring DatabaseError exception returned by renew_token so that the last request will get a AuthenticationFailed exception instead of a 500 error.

xrmx committed:
knox: handle race condition on concurrent logout call

With AUTO_REFRESH enabled authentication is racing against token
removal of logout. If a token gets removed updating its expiry
would led to a DatabaseError raised by the Django ORM.

Fix that by ignoring DatabaseError exception returned by renew_token
so that the last request will get a AuthenticationFailed exception
instead of a 500 error.

johnraz reviewed (changes requested):
Hello 👋 and thanks for the contribution!

Your branch seems to be based on master and not develop, could you please rebase on develop?

I believe the changelog would also need a note about the bug fix.

I left a comment for a potential improvement in the code as well. 

Thanks again!

xrmx commented:
@johnraz Thanks for the review, it's based on master because develop is missing some 4.1.0 commits. Will add changelog.

xrmx commented on knox/auth.py:
```diff
@@ -73,7 +74,11 @@ def authenticate_credentials(self, token):
             if compare_digest(digest, auth_token.digest):
                 if knox_settings.AUTO_REFRESH and auth_token.expiry:
-                    self.renew_token(auth_token)
+                    try:
```
The `.save` is the only database interaction in that hunk of code so there's not much chances to swallow unrelated exceptions. Regarding moving the handling inside renew_token I don't see much value in it: it will duplicate the error handling as we really can do something sensible only in authenticate_credentials.

johnraz replied:
You could swallow any other database error and mistaken it with an authentication error which would be confusing in a debug session. renew_token could change in the future and have more database interactions so I think moving it around the save makes sense. I don’t understand your argument about code duplication as this is exactly what I’m trying to avoid, using refresh token somewhere else would require to duplicate the error handling, right? Moving the error handling inside refresh token makes it safe to be used anywhere.

xrmx replied:
By duplicating error handling i mean that on top of the exception handling i have to check the return value. In my opinion it's making the code less explicit. 

```
diff --git a/knox/auth.py b/knox/auth.py
index 1d453ca..5345175 100644
--- a/knox/auth.py
+++ b/knox/auth.py
@@ -74,10 +74,8 @@ class TokenAuthentication(BaseAuthentication):
                 raise exceptions.AuthenticationFailed(msg)
             if compare_digest(digest, auth_token.digest):
                 if knox_settings.AUTO_REFRESH and auth_token.expiry:
-                    try:
-                        self.renew_token(auth_token)
-                    except DatabaseError:
-                        # avoid race condition with concurrent logout calls
+                    renewed = self.renew_token(auth_token)
+                    if not renewed:
                         continue
                 return self.validate_user(auth_token)
         raise exceptions.AuthenticationFailed(msg)
@@ -89,7 +87,12 @@ class TokenAuthentication(BaseAuthentication):
         # Throttle refreshing of token to avoid db writes
         delta = (new_expiry - current_expiry).total_seconds()
         if delta > knox_settings.MIN_REFRESH_INTERVAL:
-            auth_token.save(update_fields=('expiry',))
+            try:
+                auth_token.save(update_fields=('expiry',))
+            except DatabaseError:
+                # avoid race condition with concurrent logout calls
+                return False
+        return True
```

johnraz replied:
Yes I see your point now. I still have a code smell around this. I’m even wondering if the authentication failed error makes sense in this case? You logged in but it failed to refresh... Maybe raising a 409 (conflict) would be better, we could do a get right before the save and trigger the exception if the get fails ? That would remove the database error check I don’t like and the error checking duplication you don’t like.

xrmx replied:
It's not just that i failed to refresh, the authentication token that was valid when i started the authentication process is gone behind my back. I think authentication failure is fine in this case.

johnraz replied:
I don’t agree, you passed authentication, you are authenticated. The fact that the token is gone solely means a concurrency problem not that you failed to authenticate.

johnraz replied:
Which makes me realise that the refresh should happen at the very end of the process after validate user.

johnraz replied:
This also bring the question we discussed in #156 about getting rid of the auto refresh token feature completely, this is exposing some quirks about the design itself. Mixing authentication and refresh is troublesome.

xrmx replied:
Even if the token was valid at the time once it is gone i really think i should not consider myself authenticated. At least that's what i would like on my system.

johnraz replied:
We are circling around the same topic. This is a race condition: there is no way to know which query was triggered first, is it the authenticate call or the logout call? If the initial token check was valid, you authenticated successfully, if at the time of refresh the token is gone, this means a race condition occurred, not that you failed authentication. Deciding that you are not logged in seems way more arbitrary than telling what actually happened —> a race condition which a 409 matches. I understand it would suit you better but as a generic library we need to be as specific as possible and this is why I’m arguing here, not because I want to push something personal.

xrmx replied:
if i return a 409 and the client will retry the call with the very same token it will get a 403 no? The open question i see here is if we should make the DatabaseError in renew an error like in my version or just ignore it.

johnraz replied:
Yes a 403 should follow.

Ignoring is probably a good option If we want to be permissive and consider the request as successful.

Should the refresh failure cancel the request? If we want to be strict and follow your initial idea of considering the request as “unauthenticated behind your back”, raising is probably better.

Both solutions seem acceptable to me but I’m more on the strictness bandwagon myself.

I think doing a .get before refresh would remove the database error as I suggested above and in turn narrow the error handling to this very case. I would feel more at ease with something like this instead of a much wider database error catch.

xrmx replied:
Doing a `.get` or `refresh_from_db` would reduce the race window but not eliminate it, see TOCTOU https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use.

johnraz replied:
@xrmx this is a very valid point, thanks for the valuable link as well which helped refresh my mind on this topic 👍 

Another approach we could then take and which would probably make more sense would be to do a select for update when getting the token on login.

The rationale would be that as soon as a login is in progress, no change should be applied to a token. Basically, one can not change / destroy a token if someone is actually using it to log in but they should instead retry after the login has completed properly.

What do you think ?

xrmx replied:
I think we are overthinking this a bit :) I think in this case taking locks will create more chances to mess things than what are preventing.

johnraz replied:
Not sure, what do you have in mind that could make this worse?

johnraz replied:
oh and I just wanted to add that I agree on the overthinking bit as well, this is the feeling I have too but I don't see how to solve it simply while being "correct". Maybe we should leave it as is 😝

xrmx replied:
A worse scenario for me is code taking locks and not releasing them

johnraz commented:
If you base your work on develop and create the PR against develop, missing commits from master should not be an issue. Those are only release related commits I believe.

Writing from my phone so I may be missing a point. Maybe some other contributors could have a look. Cheers.

xrmx commented:
@johnraz yeah, but master commits should be merged back to develop or you'll have conflicts, e.g. like in this specific case in the CHANGELOG

johnraz commented:
Yes agreed on the merge back to develop but please let’s do that in a separate PR, maybe @belugame can help, won’t have access to my computer soon (travelling right now).

belugame commented on tests/tests.py:
```diff
@@ -396,3 +405,13 @@ def test_expiry_is_present(self):
+            with mock.patch.object(token_auth, 'renew_token') as m:
+                m.side_effect = DatabaseError()
+                with self.assertRaises(AuthenticationFailed):
+                    token_auth.authenticate_credentials(token.encode('utf-8'))
```
@xrmx  Thank you very much for your contribution. The scenario how you found this bug is something many could experience, however I feel your unit test is rather abstract (name and scenario). I would aim for something like test_handles_concurrent_logout_calls_as_bla_bla and try to rebuilt exactly what you experienced.

xrmx replied:
@belugame i can rename it but i cannot replicate the exact race condition deterministically inside an unit test thus the renew_token DatabaseError mock which is what i got in real life.
Please backport master 4.1.0 commit to develop so than i can rebase cleanly.

johnraz replied:
@xrmx if you provide a PR with the 4.1.0 commits in develop as I suggested before, myself or @belugame would gladly merge that.

xrmx replied:
Updated test name and added comments to describe the issue

xrmx commented:
Updated PR:
- renamed tests and added some comments to help understand the scenario
- rewritten commit message to make it hopefully more understandable
- changed the continue to a break when handling the exception as we found the token

johnraz commented:
I added the help wanted on this PR to get it rebased and merged.

@giovannicimolin if you feel like coding be my guest 😄
